### PR TITLE
step3:to7

### PR DESCRIPTION
**Pull request by LingYi0612**
## What
Some progress on Step 3 1-7

## CHECKS :warning:

Please make sure you are aware of the following.

- [ ] **The changes in this PR doesn't have private information

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -157,6 +159,8 @@ dmypy.json
 # Cython debug symbols
 cython_debug/
 
+# Database
+mercari.sqlite3
 # PyCharm
 #  JetBrains specific template is maintained in a separate JetBrains.gitignore that can
 #  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
```
Comment (karuna) [inline .gitignore:166]: also adding `.idea` might be a good here. we don't want local changes like IDE configuration to be committed accidentally.

File: python/main.py [hunk 2]
```diff
@@ -1,32 +1,170 @@
...
+
+# create table
+c.execute("""CREATE TABLE IF NOT EXISTS items (
+                id INTEGER PRIMARY KEY,
+                name STRING,
+                category STRING,
+                image STRING
+          )""")
+
+# inv = [('jacket', 'fashion'), ('cat-cage', 'pet'), ('switch', 'entertainment')]
```
Comment (karuna) [inline python/main.py:46]: usually it's considered good practice to delete unused code

https://emanuilslavov.com/remove-unused-code/

1. it doesn't add any value to the collaboration
2. it creates confusion
3. don't worry about history, if important part of the code is deleted, we can retrieve from source code history.

File: python/main.py [hunk 3]
```diff
@@ -1,32 +1,170 @@
...
+
+# Hash the image
+def hash_image(filepath):
+    with open(filepath, "rb") as f:
+        bts = f.read()  # read entire file as bytes
+        readable_hash = hashlib.sha256(bts).hexdigest();
+        return (f'{readable_hash}.jpg')
+
+
+# print(hash_image('./image/default.jpg'))
```
Comment (karuna) [inline python/main.py:74]: this is debugging code, should not be committed as well

File: python/main.py [hunk 4]
```diff
@@ -1,32 +1,170 @@
...
+
+# functions of add/get items
```
Comment (karuna) [inline python/main.py:76]: it's not adding anything, isn't?

File: python/main.py [hunk 5]
```diff
@@ -1,32 +1,170 @@
...
+        conn.close()
+        print('add succefully!')
+    except Error as e:
+        print(e)
+        return None
+
+
+# add_one_item('Kirby','game','images/default.jpg')
+# print(get_all_items())
+print(get_specific_items(name='Switch'))
```
Comment (karuna) [inline python/main.py:129]: this one is also related to code hygiene.

File: python/main.py [hunk 6]
```diff
@@ -1,32 +1,170 @@
...
+        print(e)
+        return None
+
+
+
+def get_specific_items(**args):
+    try:
+        conn = sqlite3.connect(DB_PATH)
+        c = conn.cursor()
+        if 'name' in args: c.execute("SELECT name,category,image FROM items WHERE name=?", (args['name'].lower(),))
```
Comment (karuna) [inline python/main.py:101]: this is fine for training, but consider https://xkcd.com/327/ for professional development

File: python/main.py [hunk 7]
```diff
@@ -1,32 +1,170 @@
...
+        image: str = None
+):
+    add_one_item(name, category, image)
+    result = {"name": name, "category": category, "image": image}
+    logger.info(f"Receive item: {result}")
     return {"message": f"item received: {name}"}
 
+
+@app.get("/items")
+def items():
```
Comment (karuna) [inline python/main.py:154]: why some async, some not?

File: python/main.py [hunk 8]
```diff
@@ -1,32 +1,170 @@
...
 def root():
     return {"message": "Hello, world!"}
 
+
 @app.post("/items")
-def add_item(name: str = Form(...)):
-    logger.info(f"Receive item: {name}")
+# def add_item(name: str = Form(...)):
+#     logger.info(f"Receive item: {name}")
+#     return {"message": f"item received: {name}"}
```
Comment (karuna) [inline python/main.py:141]: this one too

File: python/main.py [hunk 9]
```diff
@@ -1,32 +1,170 @@
...
+async def add_item(
+        name: str,
+        category: str,
+        image: str = None
+):
+    add_one_item(name, category, image)
```
Comment (karuna) [inline python/main.py:147]: what will happen if `add_one_item` returns None? user will see it's fine?

File: python/main.py [hunk 10]
```diff
@@ -40,3 +178,5 @@ async def get_image(items_image):
         image = images / "default.jpg"
 
     return FileResponse(image)
+
+# print(images,type(images))
```
Comment (karuna) [inline python/main.py:182]: this one too :)

File: python/main.py [hunk 11]
```diff
@@ -1,32 +1,170 @@
...
+        conn.commit()
+        conn.close()
+        print('add succefully!')
+    except Error as e:
+        print(e)
+        return None
+
+
+# add_one_item('Kirby','game','images/default.jpg')
+# print(get_all_items())
```
Comment (karuna) [inline python/main.py:128]: this one too

File: python/main.py [hunk 12]
```diff
@@ -1,32 +1,170 @@
...
+
+# functions of add/get items
+def get_all_items(json_output=False):
+    try:
+        conn = sqlite3.connect(DB_PATH)
+        c = conn.cursor()
+        sql = 'SELECT name, category, image FROM items'
+        # c.execute("""SELECT items.name,items.image,category.category_name
+        # FROM items INNER JOIN category
+        # ON items.category_id =category.category_id""")
```
Comment (karuna) [inline python/main.py:84]: this one too

File: .gitignore (hunk 1 above)
Comment (LingYi0612) [reply .gitignore:166]: so .idea is the configuration file of IDE, right?

File: python/main.py (hunk 2 above)
Comment (LingYi0612) [reply python/main.py:46]: Get it! Next time I will delete all unnecessary code before commit ;D

**Pull request closed without merging**